### PR TITLE
feat: authorization wasi hook

### PR DESCRIPTION
**Pull request by pimeys**
Just adds the wasi side of this hook. We can change the types and rename things if needed, but the machinery is in this PR.

Closes: GB-6960

### CODE REVIEW

File: engine/crates/runtime/src/user_hooks.rs [hunk 1]
```diff
@@ -51,4 +51,10 @@ pub trait UserHooksImpl: Send + Sync {
     type Context;
 
     async fn on_gateway_request(&self, headers: HeaderMap) -> Result<(Self::Context, HeaderMap), UserHookError>;
+
+    async fn on_authorization(
```
Comment (Finistere) [inline engine/crates/runtime/src/user_hooks.rs:55]: IMHO the function should be:
```
async fn authorized&self, context: &mut Self::Context, input: impl ExactSizeIterator<Item = impl serde::Serialize>) -> ...
```
Might as well use the directive name, engine will provide a &mut Self::Context. The need to use `std::mem::take` is an implementation details of wasm IMHO. No need either to specify the types of of input either. How it's serialized is also up to the implementation IMO.

Why differentiate `UserError` from `UserHookError`? In all cases it'll end up as a GraphQL error 🤷

Comment (Finistere) [reply engine/crates/runtime/src/user_hooks.rs:55]: IMHO would just `UserError` as `HookError` and that's it.

File: engine/crates/wasi-component-loader/src/hooks/authorization.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use wasmtime::{
+    component::{Resource, TypedFunc},
+    Store,
+};
+
+use crate::{names::AUTHORIZATION_HOOK_FUNCTION, state::WasiState, ComponentLoader, ContextMap, ErrorResponse};
+
+pub(crate) type HookParameters = (Resource<ContextMap>, Vec<String>);
+
+pub(crate) type HookResponse = (Result<Vec<Option<ErrorResponse>>, ErrorResponse>,);
```
Comment (Finistere) [inline engine/crates/wasi-component-loader/src/hooks/authorization.rs:12]: If those are scoped to `hooks::authorization` anyway, might as well drop the `Hook` prefix^^

**Review by Finistere — APPROVED** [review]

File: engine/crates/runtime/src/user_hooks.rs (hunk 1 above)
Comment (pimeys) [reply engine/crates/runtime/src/user_hooks.rs:55]: The `UserError` gives user control to define the message and the extensions.

Comment (Finistere) [reply engine/crates/runtime/src/user_hooks.rs]: Well yes, but in the end it doesn't make a difference to the engine whether it's wasm error or a hook error. It'll be treated the same and transformed into a GraphqlError with message & extensions. so as it'll be done in all cases we could simplify the types here IMHO. 🤷

**Pull request merged**